blog/_posts/2017-11-27-view-based-collections.md

```diff
@@ -0,0 +1,165 @@
+
+## Summary
+
+We have seen that on one hand views are useful to avoid unnecessary computations but
+the other hand they give pain to maintainers and are dangerous for end users.
+
+The collections redesign is an opportunity to fix these issues, as
+explained in the remaining of this article.
+
+## “View Based” Default Implementations
+
+Essentially, in the new collections we reversed the situation.
+Instead of having strict by default operations, they are now lazy by default and
+concrete collection types decide whether to evaluate their result or not. For instance,
+here is the implementation of `map`:
+
+~~~ scala
+trait IterableOps[A, CC[_], C] {
+  def map[B](f: A => B): CC[B] = fromIterable(View.Map(this, f))
+}
+~~~
+
+(where `CC[_]` is `List` or `Vector`)
+
+We can break down this implementation into two steps.
+
+First, we create a lazy `View` that “records” that the `map` operation has been applied.
+Constructing such a view is a cheap operation, here is the implementation of
+`View.Map`:
+
+~~~ scala
+object View {
+  case class Map[A, B](underlying: Iterable[A], f: A => B) extends View[B] {
+    def iterator() = underlying.iterator().map(f)
+  }
+}
+~~~
+
+As you can see, unless we actually iterate on the view we don’t evaluate neither the elements
+of the `underlying` collection nor the mapped elements.
+
+Second, we build the resulting collection from the view by calling `fromIterable`. This operation
+is abstract in `IterableOps` and is implemented by leaf classes of the hierarchy (e.g.
+`List`, `Vector`, etc.). Here is the implementation of `fromIterable` in `List`,
+for instance:
+
+~~~ scala
+trait List[A] extends IterableOps[A, List, List[A]] {
+  def fromIterable[E](it: Iterable[E]): List[E] = (List.newBuilder[E]() ++= it).result()
+}
+~~~
+
+This implementation creates a new list and adds all the elements of the passed `Iterable[E]`
+to it. By doing so, it necessarily computes the elements of the source collection
+and the resulting collection: `List` is a strict collection type.
+
+Now let’s have a look at the `fromIterable` implementation in `View`:
+
+~~~ scala
+trait View[A] extends IterableOps[A, View, View[A]] {
+  def fromIterable[E](it: Iterable[E]): View[E] =
+    new View[E] {
+      def iterator() = it.iterator()
+    }
+}
+~~~
+
+Here we return a view that will only evaluate its elements (by delegating to the
+`underlying` collection elements) when it is actually iterated: `View` is a lazy
+collection type.
+
+Let’s take a step back and consider again the implementation of the `map` operation:
+
+~~~ scala
+trait IterableOps[A, CC[_], C] {
+  def map[B](f: A => B): CC[B] = fromIterable(View.Map(this, f))
+}
+~~~
+
+The first part of the implementation creates a lazy view of the transformed elements,
+and the second part of the implementation (the call to `fromIterable`) turns this view
+into the resulting collection (which can be done either in a lazy or strict way, as
+we’ve just seen).
```

Comment:
I'd mention that this is not the actual implementation of `map` in `List`. We override it specifically for better performance in `List` and we also have a more generic strict version available as a mix-in for other strict collection types because it is faster than the potentially lazy default implementation.

Reply:
I plan to write another article specifically about the consequences on performance and how we fixed it.